typing/ctype.ml

```diff
@@ -1831,20 +1831,31 @@ let is_contractive env p =
 
 exception Occur
 
+(** [occur_rec ...allow_recursive ... ty0 ty] checks if the node of [ty0] is
+    reachable from the node [ty] follwing a path which is not guarded by allowed
+    recursive constructs. More precisely, recursion under polymorphic variant
+    and object types is always allowed and if [allow_recursive=true] recursion
+    under contractive constructor is also allowed. *)
 let rec occur_rec env allow_recursive visited ty0 ty =
   if eq_type ty ty0 then raise Occur;
   match get_desc ty with
     Tconstr(p, _tl, _abbrev) ->
       if allow_recursive && is_contractive env p then () else
-      if TypeSet.mem ty visited then
-        if allow_recursive then
-          retry_after_expansion env allow_recursive visited ty0 ty
-        else raise Occur
+      if TypeSet.mem ty visited then raise Occur
```

Comment:
We discussed this PR together and I remember some of the details, but I still find this line puzzling -- despite the fact that it was already there before. If `ty` is a type for which we already tried expansion and it failed, then I understand that we want to reject here. But with the code as written, `ty` is already marked as visited in the first step, when we check its arguments without expanding, and there turning recursive occurrences of `ty` into "I detected an occurrence of `ty0`" feels wrong. What assumptions are we making to know that this case cannot happen, or would already have been rejected somewhere else?

Reply:
More precisely, I would expect the following:
1. there is a set of "safe" types and a set of "unsafe" types
  (recursive occurrences of safe types stop the search,
   recursive occurrences of unsafe types are considered as an occurrence)
2. when we go under a non-abbreviation type-former, we mark the type node as safe
3. when we visit the arguments of an expansion, we mark the type node as safe
4. when we expand the body of an expansion, we mark the type node as unsafe

It may be that some other invariant verified by the type-checker (for example that expansions are not ill-founded in the sense of typedecl.ml:check_well_founded_*) implies that point (3) is moot, because there will never be a recursive occurrence of the type itself ungarded in its arguments. If that was the case, then the logic above would be correct, despite its suspicious looks. (It corresponds to marking the type node as *unsafe* at step (3), which is equivalent to marking it safe if we know there are no recursive occurrences in this part of the graph anyway.)